compiler/rustc_mir_dataflow/src/impls/liveness.rs

```diff
@@ -137,23 +138,23 @@ impl DefUse {
         match context {
             PlaceContext::NonUse(_) => None,
 
-            PlaceContext::MutatingUse(MutatingUseContext::Store | MutatingUseContext::Deinit) => {
-                Some(DefUse::Def)
-            }
+            PlaceContext::MutatingUse(
+                MutatingUseContext::Store
+                | MutatingUseContext::Deinit
+                | MutatingUseContext::AsmOutput,
+            ) => Some(DefUse::Def),
 
             // Setting the discriminant is not a use because it does no reading, but it is also not
             // a def because it does not overwrite the whole place
-            PlaceContext::MutatingUse(MutatingUseContext::SetDiscriminant) => None,
+            PlaceContext::MutatingUse(MutatingUseContext::SetDiscriminant) => {
+                Some(DefUse::MaybeDef)
+            }
```

Comment:
Is this introduced to ensure that an indirect use of a base local in `SetDiscriminant` is considered to be a use of that local? For example in `discriminant(*_1) = 0`? 

I think `call_return_effect` and `yield_resume_effect` have analogous issue (note that the box allocation was optimized out):

```rust
#![feature(box_syntax)]
#![feature(core_intrinsics)]

fn main() {
    let a: Box<[u32; 20]> = box Default::default();
    core::intrinsics::forget(a);
}
```

```console
$ rustc +1868626d2fe06c472a30fa1bd21b31f154c681e2 a.rs -O -Cpanic=abort -Zunpretty=mir

    ...

    bb1: {
        StorageLive(_5);                 // scope 0 at a.rs:5:29: 5:51
        (*_5) = <[u32; 20] as Default>::default() -> bb2; // scope 0 at a.rs:5:33: 5:51
                                         // mir::Constant
                                         // + span: a.rs:5:33: 5:49
                                         // + literal: Const { ty: fn() -> [u32; 20] {<[u32; 20] as Default>::default}, val: Value(Scalar(<ZST>)) }
```